src/main/scala/emm/effects/Binder.scala

```diff
@@ -29,6 +30,22 @@ trait BinderLowPriorityImplicits {
       NAP.pack(back)
     }
   }
+
+  implicit def binderState[S, C <: Effects](implicit C: Binder[C], T: Traverser[C], NN: NonNested[C]): Binder[State[S, ?] |: C] = new Binder[State[S, ?] |: C] {
```

Comment:
> On the flip side, the fact that we may need to support some monad transformer WhateverT has me questioning if Emm is general enough. The problem with State is the inner F in F[S => F[(S, A)]]. We're not just wrapping State in some monad. We're also wrapping the return of the function, namely F[(S, A)]. This turns it into a special case, which I don't like, because I don't like special cases...

I completely agree with this.  Basically, at a fundamental level, `StateT` is a different algebra from `State`.  You can define `State` in terms of `StateT`, but the inverse is not true.  This is different from something like `ListT` or `OptionT`, where it is relatively easy to define one in terms of the other.  This is a critical distinction for `Emm`, where we are literally deriving (at compile time) a transformer for each monad in the stack defined in terms of the monad itself, which is impossible with `State`.  I would dance for joy if we could get rid of the special casing or (equally importantly) not have to depend on the existence of transformers like `StateT`.  I just can't see a way.

> I need to go look at implementations, but thought: why does it make sense for the run of StateT to have F in both places. Namely, would F[S => G[(S, A)]] be (a) possible and (b) useful?

Well, `StateT` has the outer context for broadly no reason at all.

```scala
def proof[F[_]: Monad, S, A](fs: F[S => F[(S, A)]])(s: S): F[(S, A)] = Functor[F].flatMap(fs) { _(s) }
```

In words… the outer context does not gain us (well, `StateT`) anything, since it is trivial to flatten the outer context into the inner provided the outer context has a `Monad`, which is already a constraint on several other functions.

As to having a different inner context from the outer…  That could potentially lead to a variant of `StateT` which changes its transformed effect between stages of the computation (similar to how `IndexedStateT` can change its state type between stages).  There would probably need to be some constraint though, since I think evaluation would require broadening one of the effects into the other (more generally: we would need to satisfy the type `exists H[_] . (F ~> H, G ~> H)`).  Not sure this helps `Emm` at all, but it's a cool thought.

> Another realization. StateT is almost just wrapping Kleisli in the same F[_] the Kleisli returns. The difference being, the tuple in the return type. But that's just a specific case that we might not need to match on individually. Meaning, perhaps we could match StateT on something like F |: Kleisli |: Base.

Thought about that.  I am relatively certain that `StateT` can indeed be defined in terms of `Kleisli`, though I haven't actually tried to do it.  The utility functions that `StateT` provides are the useful bit though (e.g. `get`, `set`, etc), and those cannot be automagically derived.